Jump to definition of commands/environments

mttnff opened the pull request:
Similar to references and labels, this PR adds the functionality to quickly jump from a command or environment name to the corresponding definition. This feature has been requested before (on [sourceforge](https://sourceforge.net/p/texstudio/feature-requests/628/)) and I find it quite useful in long documents. The behaviour implemented in this PR is just a proposal which I'm happy to discuss and improve.

Implementation details:
 * Definitions of commands/environments are tracked in `LatexDocument` using the new member `mUserCommandGoto` (the existing `mUserCommandList` provides similar information for completion, one could consider merging the two).
 * New `LinkOverlay`s are added for commands and environment names (tokens `command`, `unknownCommand`, `beginEnv`, `env`). An exception are the commands `\begin` and `\end`, for which no link overlay is added.
 * The added implementation in `Texstudio::editGotoDefinition` is analogous to (actually, mostly copied from) the code for references/labels.

    The last commit in the PR additionally adds the functionality to jump to `\usepackage` for known commands/environments (if no user definition is found). An exception are the default packages "tex", "latex-document" and "latex-mathsymbols"; for commands from these packages, no jump is performed (are there more such packages that are included by default?). I'm not sure how useful jumping to `\usepackage` really is, so this last commit is rather optional.

Further changes and remarks:
 * The `LinkOverlay` constructor now accepts a `Token` to determine the position and length of the overlay, instead of manually searching for brackets. I think this might be more consistent with the remaining code, but I'm not completely sure that it results in the same overlays in all cases.
 * The context menu on `\begin{...}` provides information about the package of the environment (if known). The same information is now included if the menu is opened on the name of the environment (i.e., on `...`). The intention is to be more consistent with the "jump-to-definition" feature that currently only works when clicking on the name (see image below).
 * For an environment `env`, the commands `\env` and `\endenv` are added to the new `mUserCommandGoto`, but are otherwise not considered by TeXstudio. If we want to support them for "jump-to-definition", we might also want to add them to `mUserCommandList` for code completion.

The new functionality in action:
![env_link](https://user-images.githubusercontent.com/15823528/67162733-4701eb80-f367-11e9-8fe3-71163d2f5df8.png)
![Screenshot_20191020_182916](https://user-images.githubusercontent.com/15823528/67162752-89c3c380-f367-11e9-9242-42dba7c0e44e.png)

sunderme commented on src/latexdocument.cpp:
```diff
@@ -397,6 +400,20 @@ inline bool isDefinitionArgument(const QString &arg)
+ * \param name name of the command (or environment)
+ * \param start column at which the definition begins
+ */
+void LatexDocument::addUserCommandDefinition(const QDocumentLine line, const QString &name, int start) {
```
QDocumentLine as argument seems to be not very useful here as you only use the linehandle.

mttnff replied:
Indeed, updated it to only take the line handle.

MeanSquaredError commented:
The PR seems to build OK, but Travis CI has timed out again while updating a gpg key, pretty much like it did with my patch a couple of days ago.

So rerunning the Travis CI checks will probably help to remove the error like it did with my PR.

mttnff commented on src/texstudio.cpp:
```diff
@@ -1892,6 +1892,29 @@ LatexEditorView *Texstudio::editorViewForLabel(LatexDocument *doc, const QString
 }
 
+LatexEditorView *Texstudio::editorViewForCommandDefinition(LatexDocument *doc, const QString &command)
+{
```
Not sure if I understand your comment. These methods have to look up the definition of the given label/command/package in order to determine the corresponding editor view. I agree that there is some code duplication here, but I don't see how using tokens would help.

sunderme replied:
the idea is that if you have basically the same method with only one line changed, you could use one method with a if/them selection on where to look for the needed info. And since the token tells you, it is a command, package,reference, one could just pass the token and let the method figure out the rest...

mttnff replied:
I see your point. The case distinction on the token is already done in `Texstudio::editGotoDefinition` and I would prefer to have it in one place only. So I would propose to move the lookup to `editGotoDefinition` and then simply remove these methods (since they do not seem to be used from anywhere else).

If we do the same for `gotoToLabel` and `gotoToCommand` etc., we can also avoid the duplicate lookup (currently once to find the editor view and then again to jump to the correct line).

sunderme replied:
Sounds reasonable to me.

mttnff commented:
Thanks for your feedback. I'm not very familiar with TeXstudio's implementation yet, so there is most probably room for improvement. I have answered two of your comments above.

The reason why I added the extra storage `mUserCommandGoto` is that it serves a slightly different purpose than `mUserCommandList`. As far as I understand, the latter is used for completion and stores a `CodeSnippet`, whereas `mUserCommandGoto` only stores the names of commands/environments to look up their definition.

I guess it would be possible to rely on the information provided by `CodeSnippet` in the implementation of `LatexDocument::getCommandDefinitions` so that we do not need the extra storage. Since `CodeSnippet` is only intended for completion, I'm not sure if this would be a reasonable thing to do?

mttnff commented on src/latexdocument.cpp:
```diff
@@ -1412,6 +1440,35 @@ QMultiHash<QDocumentLineHandle *, int> LatexDocument::getLabels(const QString &n
+	return result;
+}
+
+QSet<QDocumentLineHandle *> LatexDocument::getUsePackages(const QString &name)
```
I tried to mimic the signatures of `getLabels` and `getBibItems`; will change it to a list.

mttnff replied:
This is part of the refactoring; these methods now simply return the first matching line handle.

mttnff commented on src/latexdocument.cpp:
```diff
@@ -1412,6 +1440,35 @@ QMultiHash<QDocumentLineHandle *, int> LatexDocument::getLabels(const QString &n
+			if (rp.name == name && elem->indexOf(it.key()) >= 0) {
+				result.insert(it.key(), rp.start);
+			}
+		}
```
Here we could instead compare `name` to the `CodeSnippet`s stored in `mUserCommandList`, in order to get rid of `mUserCommandGoto`.

sunderme replied:
I would prefer usind mUserCommandList for all functionality. It might need extensions, but i am not a fan of redundant structures which are forgotten in the next refactorizatiom

mttnff replied:
Okay, I will look into this. I think storing a struct containing both the name (for goto definition) and the CodeSnippet (for completion) would be a good idea then.

mttnff replied:
This is part of the refactoring.

mttnff commented:
I have done some refactoring according to our discussion. Details:

 * `mUserCommandList` now stores entries of type `UserCommandPair` which contain both the code snippet for code completion (as before) and a string to store the name of commands and environments. I removed the redundant `mUserCommandGoto`.

 * The method `Texstudio::editGotoDefinition` now does most of the work and calls `getLabels` etc. directly in order to look up definitions. I have thus removed/shortened the methods `editorViewForLabel` as well as `gotoToLabel` (the latter is now called `gotoLineHandleAndSearchLabel` to match the naming of similar methods).

 * The lookup methods `getUsePackages` and `getCommandDefinition` now return the first matching line handle, instead of a set/list of all matches. I have renamed them to `findUsePackage` and `findCommandDefinition` to reflect this change. They are currently only used by this PR, so this does not affect other code.

Questions:

 * For special definitions such as newcolumntype or definecolor, the `name` field in `UserCommandPair` is currently set to a null string. It might make sense to also store some kind of name in these cases, but I'm not sure what to use (this should not be relevant for the "goto definition"-feature).

There are only minor conflicts with the master branch which can be resolved by a rebase.

sunderme commented:
thanks.
It would probably be nice to have automated tests on this ...

sunderme commented:
context menu on commands where jump to definition works, would be nice ...
